stdlib/random.ml

```diff
@@ -13,76 +14,69 @@
-   function to enhance the mixing of bits.
-   If we use normal addition, the low-order bit fails tests 1 and 7
-   of the Diehard test suite, and bits 1 and 2 also fail test 7.
-   If we use multiplication as suggested by Marsaglia, it doesn't fare
-   much better.
-   By mixing the bits of one of the numbers before addition (XOR the
-   5 high-order bits into the low-order bits), we get a generator that
-   passes all the Diehard tests.
-*)
+(* Pseudo-random number generator *)
 
 external random_seed: unit -> int array = "caml_sys_random_seed"
 
 module State = struct
 
-  type t = { st : int array; mutable idx : int }
+  open Bigarray
 
-  let new_state () = { st = Array.make 55 0; idx = 0 }
-  let assign st1 st2 =
-    Array.blit st2.st 0 st1.st 0 55;
-    st1.idx <- st2.idx
+  type t = (int64, int64_elt, c_layout) Array1.t
 
+  external next: t -> (int64[@unboxed])
+      = "caml_lxm_next" "caml_lxm_next_unboxed" [@@noalloc]
 
-  let full_init s seed =
-    let combine accu x = Digest.string (accu ^ Int.to_string x) in
-    let extract d =
-      Char.code d.[0] + (Char.code d.[1] lsl 8) + (Char.code d.[2] lsl 16)
-      + (Char.code d.[3] lsl 24)
-    in
-    let seed = if Array.length seed = 0 then [| 0 |] else seed in
-    let l = Array.length seed in
-    for i = 0 to 54 do
-      s.st.(i) <- i;
-    done;
-    let accu = ref "x" in
-    for i = 0 to 54 + Int.max 55 l do
-      let j = i mod 55 in
-      let k = i mod l in
-      accu := combine !accu seed.(k);
-      s.st.(j) <- (s.st.(j) lxor extract !accu) land 0x3FFFFFFF;  (* PR#5575 *)
-    done;
-    s.idx <- 0
+  let create () : t =
+    Array1.create Int64 C_layout 4
 
+  let set s i1 i2 i3 i4 =
+    Array1.unsafe_set s 0 (Int64.logor i1 1L); (* must be odd *)
+    Array1.unsafe_set s 1 i2;
+    Array1.unsafe_set s 2 (if i3 <> 0L then i3 else 1L); (* must not be 0 *)
+    Array1.unsafe_set s 3 (if i4 <> 0L then i4 else 2L) (* must not be 0 *)
 
-  let make seed =
-    let result = new_state () in
-    full_init result seed;
-    result
-
+  let mk i1 i2 i3 i4 =
+    let s = create () in
+    set s i1 i2 i3 i4; s
 
-  let make_self_init () = make (random_seed ())
+  let assign (dst: t) (src: t) =
+    Array1.blit src dst
 
   let copy s =
-    let result = new_state () in
-    assign result s;
-    result
+    let s' = create() in assign s' s; s'
+
+  (* The seed is an array of integers.  It can be just one integer,
+     but it can also be 12 or more bytes.  To hide the difference,
+     we serialize the array as a sequence of bytes, then hash the
+     sequence with MD5 (Digest.bytes).  MD5 gives only 128 bits while
+     we need 256 bits, so we hash twice with different suffixes. *)
+  let reinit s seed =
+    let n = Array.length seed in
+    let b = Bytes.create (n * 8 + 1) in
+    for i = 0 to n-1 do
+      Bytes.set_int64_le b (i * 8) (Int64.of_int seed.(i))
+    done;
+    Bytes.set b (n * 8) '\x01';
+    let d1 = Bytes.unsafe_of_string (Digest.bytes b) in
```

Comment:
No need to use Bytes.unsafe_of_string if we use String.get_int64_le below

Reply:
Very good point, thanks!  I was not aware that `get_*` functions had been added to `String`...